File: sirius-core/src/main/java/org/starcoin/sirius/core/TransferDeliveryChallenge.java
package org.starcoin.sirius.core;

import org.starcoin.core.MerklePath;

import java.util.logging.Logger;

public class TransferDeliveryChallenge {

    private Update update;

    private ChallengeStatus status;

    private OffchainTransaction transaction;

    private MerklePath<OffchainTransaction> provePath;

    private Logger logger = Logger.getLogger(TransferDeliveryChallenge.class.getName());

    public TransferDeliveryChallenge(Update update, OffchainTransaction trans, MerklePath path) {
        this.update = update;
        this.transaction = trans;
        this.provePath = path;
    }

    public Update getUpdate() {
        return this.update;
    }

    public void setUpdate(Update update) {
        this.update = update;
    }

    public ChallengeStatus getStatus() {
        return this.status;
    }

    public void setStatus(ChallengeStatus status) {
        this.status = status;
    }

    public OffchainTransaction getTransaction() {
        return this.transaction;
    }

    public void setTransaction(OffchainTransaction transaction) {
        this.transaction = transaction;
    }

    public MerklePath getProvePath() {
        return this.provePath;
    }

    public void setProvePath(MerklePath provePath) {
        this.provePath = provePath;
    }

    public boolean openChallenge() {
        synchronized (this) {
            if (this.status == null) {
                this.status = ChallengeStatus.OPEN;
                logger.warning("openChallenge succ");
                return true;
            }

            logger.warning(
                    "openChallenge err status : " + ((this.status == null) ? "null" : this.status));
            return false;
        }
    }

    public boolean closeChallenge() {
        synchronized (this) {
            if (this.status != null && this.status == ChallengeStatus.OPEN) {
                this.status = ChallengeStatus.CLOSE;
                logger.warning("closeChallenge succ");
                return true;
            }

            logger.warning(
                    "openChallenge err status : " + ((this.status == null) ? "null" : this.status));
            return false;
        }
    }

    public boolean isClosed() {
        synchronized (this) {
            return (this.status != null && this.status == ChallengeStatus.CLOSE);
        }
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/TransferDeliveryChallenge.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema

@ProtobufSchema(Starcoin.TransferDeliveryChallenge::class)
@Serializable
data class TransferDeliveryChallenge(
    @SerialId(1)
    val update: Update = Update.DUMMY_UPDATE,
    @SerialId(2)
    val tx: OffchainTransaction = OffchainTransaction.DUMMY_OFFCHAIN_TRAN,
    @SerialId(3)
    val path: MerklePath = MerklePath.DUMMY_PATH
) : SiriusObject() {
    companion object : SiriusObjectCompanion<TransferDeliveryChallenge, Starcoin.TransferDeliveryChallenge>(
        TransferDeliveryChallenge::class
    ) {

        var DUMMY_TRAN_DELIVERY_CHALLENGE = TransferDeliveryChallenge()

        override fun mock(): TransferDeliveryChallenge {
            return TransferDeliveryChallenge(Update.mock(), OffchainTransaction.mock(), MerklePath.mock())
        }
    }
}

@ProtobufSchema(Starcoin.TransferDeliveryChallengeStatus::class)
@Serializable
data class TransferDeliveryChallengeStatus(
    @SerialId(1) val challenge: TransferDeliveryChallenge = TransferDeliveryChallenge.DUMMY_TRAN_DELIVERY_CHALLENGE,
    @SerialId(2) val status: ChallengeStatus = ChallengeStatus.OPEN
) :
    SiriusObject() {
    companion object : SiriusObjectCompanion<TransferDeliveryChallengeStatus, Starcoin.TransferDeliveryChallengeStatus>(
        TransferDeliveryChallengeStatus::class
    ) {

        var DUMMY_TRAN_DELIVERY_CHALLENGE = TransferDeliveryChallengeStatus()

        override fun mock(): TransferDeliveryChallengeStatus {
            return TransferDeliveryChallengeStatus(TransferDeliveryChallenge.mock(), ChallengeStatus.OPEN)
        }
    }
}
